File: zypher-crm/frontend/src/types/deal.ts
// frontend/src/types/deal.ts - Tipos TypeScript para Deals (Phase 3)

// Enums para Deal
export type DealStage = 'lead' | 'contacted' | 'qualified' | 'proposal' | 'negotiation' | 'won' | 'lost'

export type DealPriority = 'low' | 'medium' | 'high' | 'urgent'

export type DealSource =
  | 'website'
  | 'referral'
  | 'cold_call'
  | 'social_media'
  | 'email_campaign'
  | 'event'
  | 'partner'
  | 'other'

// Interface base para Deal
export interface DealBase {
  title: string
  description: string | null
  value: number
  currency: string
  stage: DealStage
  probability: number
  priority: DealPriority
  source: DealSource | null
  expected_close_date: string | null
  actual_close_date: string | null
  lost_reason: string | null
  tags: string[]
  custom_fields: Record<string, any>
  is_active: boolean
}

// Interface para crear Deal
export interface DealCreate extends DealBase {
  contact_id: string
}

// Interface para actualizar Deal
export interface DealUpdate extends Partial<DealBase> {}

// Interface para Deal completo (respuesta del servidor)
export interface DealOut extends DealBase {
  id: string
  contact_id: string
  owner_id: number
  created_at: string
  updated_at: string
}

// Interface para Deal con información del contacto
export interface DealWithContact extends DealOut {
  contact_name: string | null
  contact_email: string | null
  contact_company: string | null
}

// Interface para respuesta paginada
export interface DealsPaginatedResponse {
  items: DealOut[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// Interface para estadísticas de deals
export interface DealStats {
  total_deals: number
  active_deals: number
  won_deals: number
  lost_deals: number
  total_value: number
  won_value: number
  potential_value: number
  average_deal_value: number
  average_win_probability: number
  deals_by_stage: Record<DealStage, number>
  value_by_stage: Record<DealStage, number>
  deals_by_priority: Record<DealPriority, number>
  deals_by_source: Record<DealSource, number>
  win_rate: number
  conversion_rate: number
}

// Interface para actualizar etapa
export interface DealStageUpdate {
  stage: DealStage
  probability?: number
  lost_reason?: string
}

// Interface para resumen del pipeline
export interface PipelineStageData {
  stage: DealStage
  count: number
  total_value: number
  deals: DealOut[]
}

export type PipelineSummary = Record<DealStage, PipelineStageData>

// Labels para las etapas
export const StageLabels: Record<DealStage, string> = {
  lead: 'Lead',
  contacted: 'Contactado',
  qualified: 'Calificado',
  proposal: 'Propuesta',
  negotiation: 'Negociación',
  won: 'Ganado',
  lost: 'Perdido',
}

// Labels para prioridades
export const PriorityLabels: Record<DealPriority, string> = {
  low: 'Baja',
  medium: 'Media',
  high: 'Alta',
  urgent: 'Urgente',
}

// Labels para fuentes
export const SourceLabels: Record<DealSource, string> = {
  website: 'Sitio Web',
  referral: 'Referido',
  cold_call: 'Llamada en Frío',
  social_media: 'Redes Sociales',
  email_campaign: 'Campaña Email',
  event: 'Evento',
  partner: 'Socio/Partner',
  other: 'Otro',
}

// Opciones para selects
export const STAGE_OPTIONS = Object.entries(StageLabels).map(([value, label]) => ({
  value: value as DealStage,
  label,
}))

export const PRIORITY_OPTIONS = Object.entries(PriorityLabels).map(([value, label]) => ({
  value: value as DealPriority,
  label,
}))

export const SOURCE_OPTIONS = Object.entries(SourceLabels).map(([value, label]) => ({
  value: value as DealSource,
  label,
}))

// Colores para etapas (para UI)
export const StageColors: Record<DealStage, string> = {
  lead: 'gray',
  contacted: 'blue',
  qualified: 'cyan',
  proposal: 'purple',
  negotiation: 'orange',
  won: 'green',
  lost: 'red',
}

// Colores para prioridades
export const PriorityColors: Record<DealPriority, string> = {
  low: 'gray',
  medium: 'blue',
  high: 'orange',
  urgent: 'red',
}

// Probabilidades sugeridas por etapa
export const StageProbabilities: Record<DealStage, number> = {
  lead: 10,
  contacted: 20,
  qualified: 40,
  proposal: 60,
  negotiation: 80,
  won: 100,
  lost: 0,
}

File: backend/app/models/deal.py
from sqlalchemy import Column, String, Text, ForeignKey, Float, Date, Enum as SQLEnum, JSON, Boolean, Integer
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
import enum
import uuid

from app.models.base import BaseModel


class DealStage(str, enum.Enum):
    """Etapas del pipeline de ventas"""
    LEAD = "LEAD"
    CONTACTED = "CONTACTED"
    QUALIFIED = "QUALIFIED"
    PROPOSAL = "PROPOSAL"
    NEGOTIATION = "NEGOTIATION"
    WON = "WON"
    LOST = "LOST"


class DealPriority(str, enum.Enum):
    """Prioridad del deal"""
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    URGENT = "URGENT"


class DealSource(str, enum.Enum):
    """Fuente de origen del deal"""
    WEBSITE = "WEBSITE"
    REFERRAL = "REFERRAL"
    COLD_CALL = "COLD_CALL"
    SOCIAL_MEDIA = "SOCIAL_MEDIA"
    EMAIL_CAMPAIGN = "EMAIL_CAMPAIGN"
    EVENT = "EVENT"
    PARTNER = "PARTNER"
    OTHER = "OTHER"


class Deal(BaseModel):
    """Modelo de negocio/oportunidad en el pipeline - PHASE 3"""
    __tablename__ = "deals"

    # Cambiamos a UUID para mejor escalabilidad
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)

    # Información básica
    title = Column(String(300), nullable=False, index=True)
    description = Column(Text, nullable=True)

    # Valor del negocio
    value = Column(Float, default=0.0, nullable=False, index=True)
    currency = Column(String(10), default="USD", nullable=False)

    # Etapa en el pipeline
    stage = Column(SQLEnum(DealStage), default=DealStage.LEAD, nullable=False, index=True)

    # Probabilidad de cierre (0-100)
    probability = Column(Integer, default=0, nullable=False)

    # Prioridad del deal
    priority = Column(SQLEnum(DealPriority), default=DealPriority.MEDIUM, nullable=False)

    # Fuente de origen
    source = Column(SQLEnum(DealSource), default=DealSource.OTHER, nullable=True)

    # Fechas importantes
    expected_close_date = Column(Date, nullable=True, index=True)
    actual_close_date = Column(Date, nullable=True)

    # Motivo de pérdida (si stage = LOST)
    lost_reason = Column(String(500), nullable=True)

    # Tags para categorización
    tags = Column(JSON, nullable=True, default=list)

    # Campos personalizados (JSON flexible)
    custom_fields = Column(JSON, nullable=True, default=dict)

    # Estado activo
    is_active = Column(Boolean, default=True, nullable=False, index=True)

    # Relaciones (UUID en lugar de Integer)
    contact_id = Column(UUID(as_uuid=True), ForeignKey("contacts.id"), nullable=False, index=True)
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)

    # Relationships (comentadas por ahora, activar cuando se implementen back_populates)
    # contact = relationship("Contact", back_populates="deals")
    # owner = relationship("User", back_populates="deals")
    # tasks = relationship("Task", back_populates="deal", cascade="all, delete-orphan")
    # notes = relationship("Note", back_populates="deal", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Deal {self.title} - {self.stage} - ${self.value}>"
